### ISSUE TITLE
feature: Conditional progress bar based on log level

### ISSUE THREAD

**Issue by frereit** [athre0z/zzz#3]
I have the following use case:

I want to display the progress bar only if the log level is set to "INFO" or higher (as specified by the [log crate](https://docs.rs/log/latest/log/#in-libraries)). 

Currently, I implement the following trait manually inside my crate:

```rust
use log::{log_enabled, Level};
use zzz::{ProgressBarIterExt, ProgressBarIter};


// Helper type to hold the iterator, with a progress bar if desired
pub(crate) enum OptionalProgressBarIter<I: Iterator> {
    Enabled(ProgressBarIter<I>),
    Disabled(I),
}

impl<I: Iterator> Iterator for OptionalProgressBarIter<I> {
    type Item = I::Item;

    fn next(&mut self) -> Option<Self::Item> {
        // Simply iterate from the underlying iterator
        match self {
            OptionalProgressBarIter::Enabled(iter) => iter.next(),
            OptionalProgressBarIter::Disabled(iter) => iter.next(),
        }
    }
}

pub(crate) trait ProgressBarIterLogExt: Iterator {
    fn progress_with_level<'a>(self, level: Level) -> OptionalProgressBarIter<Self>
    where
        Self: Sized + 'a,
    {
        if log_enabled!(level) {
            OptionalProgressBarIter::Enabled(self.progress())
        } else {
            OptionalProgressBarIter::Disabled(self)
        }
    }
}
impl<I> ProgressBarIterLogExt for I where I: Iterator {}
```

This API can then be used like this:

```rust
some_iterator.progress_with_level(Level::Info)
```

Is this API something that you would consider adding into your crate? I think this is a common enough use case to warrant it but it's up to you.

**Comment by athre0z** [comment]
Hmm, I'm not fully convinced that this particular use case is common enough to warrant special-casing in general. Personally I mostly stopped using the `log` crate in favor of `tokio-tracing`. I'd suggest a more generalized approach of either:

1. adding a `should_render Option<fn() -> bool>` (or `Option<Box<dyn FnMut() -> bool>>`) to `ProgressBarConfig`. Users could then specify whatever custom logic to decide whether the progress bar should render or not and just pass in their config instance wherever needed. The drawback here is that the convenient `.progress()` extension cannot be used and you'd always have to use `.with_progress(ProgressBar::smart().config(&MY_CONFIG))`, the upside is that the choice would be per call site. Adding a `.progress_with_config(..)` method would also be an option.

1. add a global static with a similar callback + a public setter method that can be used to customize this behavior globally. No ability to customize the behavior per call-site, but very convenient and consistent behavior across the application -- somewhat consistent with how the log crate has the log level as a global static

1. do both, with the callback in the config taking precedence over the global if it is set

I think my preference here would be to go with option 2 and then perhaps expand to option 3 if need arises later. What do you think?

**Comment by frereit** [comment]
> go with option 1 and just allow users to replace the global default config. That would kind of achieve the same thing as option 2 without additional complexity, so I think this might actually be my new preference.
I agree, this seems like the most ergonomic option.

**Comment by athre0z** [comment]
Implemented in 6c3fd37ef8df36c2e81a86e4c4656c1582f7c847 and 6c3fd37ef8df36c2e81a86e4c4656c1582f7c847 and published as part of version `v0.3.0`.
